File: src/main/java/architecture/ego_equipment/datagen/Datagen.java
package architecture.ego_equipment.datagen;

import architecture.ego_equipment.core.EGOEquipment;
import architecture.ego_equipment.datagen.i18n.ZhCn;
import architecture.ego_equipment.datagen.tag.DatagenBlockTag;
import architecture.ego_equipment.datagen.tag.DatagenItemTag;
import net.minecraft.core.HolderLookup;
import net.minecraft.core.RegistrySetBuilder;
import net.minecraft.data.DataGenerator;
import net.minecraft.data.DataProvider;
import net.minecraft.data.PackOutput;
import net.neoforged.bus.api.SubscribeEvent;
import net.neoforged.fml.common.EventBusSubscriber;
import net.neoforged.neoforge.common.data.ExistingFileHelper;
import net.neoforged.neoforge.data.event.GatherDataEvent;
import org.jetbrains.annotations.NotNull;

import java.util.concurrent.CompletableFuture;

/**
 * 数据生成主类
 */
@EventBusSubscriber(modid = EGOEquipment.ID)
public final class Datagen {
	@SubscribeEvent
	public static void gatherData(@NotNull GatherDataEvent event) {
		DataGenerator generator = event.getGenerator();
		PackOutput output = generator.getPackOutput();
		CompletableFuture<HolderLookup.Provider> completableFuture = event.getLookupProvider();

		ExistingFileHelper existingFileHelper = event.getExistingFileHelper();
		// 服务端数据生成
		buildServer(event, generator, new ModDatagenDatapackBuiltinEntries(output, completableFuture, new RegistrySetBuilder()));

		DatagenBlockTag blockTag = new DatagenBlockTag(output, completableFuture, existingFileHelper);
		buildServer(event, generator, blockTag);
		buildServer(event, generator, new DatagenItemTag(output, completableFuture, blockTag.contentsGetter(), existingFileHelper));

		// 客户端数据生成
		buildClient(event, generator, new DatagenParticle(output, existingFileHelper));
		buildClient(event, generator, new ZhCn(output));
		buildClient(event, generator, new DatagenItemModel(output, existingFileHelper));
	}

	private static <T extends DataProvider> @NotNull T buildClient(
		@NotNull GatherDataEvent event,
		@NotNull DataGenerator generator,
		T provider
	) {
		return generator.addProvider(event.includeClient(), provider);
	}

	private static <T extends DataProvider> @NotNull T buildServer(
		@NotNull GatherDataEvent event,
		@NotNull DataGenerator generator,
		T provider
	) {
		return generator.addProvider(event.includeServer(), provider);
	}
}

File: src/main/kotlin/architecture/ego_equipment/datagen/Datagen.kt
package architecture.ego_equipment.datagen

import architecture.ego_equipment.core.EGOEquipment
import architecture.ego_equipment.datagen.i18n.ZhCn
import architecture.ego_equipment.datagen.tag.DatagenBlockTag
import architecture.ego_equipment.datagen.tag.DatagenItemTag
import architecture.goldenboughs_lib.util.buildClient
import architecture.goldenboughs_lib.util.buildServer
import net.minecraft.core.RegistrySetBuilder
import net.minecraft.data.DataGenerator
import net.minecraft.data.DataProvider
import net.neoforged.bus.api.SubscribeEvent
import net.neoforged.fml.common.EventBusSubscriber
import net.neoforged.neoforge.data.event.GatherDataEvent

@EventBusSubscriber(modid = EGOEquipment.ID)
object Datagen {
	@SubscribeEvent
	fun gatherData(event: GatherDataEvent) {
		val generator = event.generator
		val output = generator.packOutput
		val completableFuture = event.lookupProvider
		val existingFileHelper = event.existingFileHelper

		event.buildServer(ModDatagenDatapackBuiltinEntries(output, completableFuture, RegistrySetBuilder()))

		val blockTag = DatagenBlockTag(output, completableFuture, existingFileHelper)
		event.buildServer(blockTag)
		buildServer(
			event,
			generator,
			DatagenItemTag(output, completableFuture, blockTag.contentsGetter(), existingFileHelper)
		)

		event.buildClient(DatagenParticle(output, existingFileHelper))
		event.buildClient(ZhCn(output))
		event.buildClient(DatagenItemModel(output, existingFileHelper))
	}

	private fun <T : DataProvider> buildClient(event: GatherDataEvent, generator: DataGenerator, provider: T): T =
		generator.addProvider(event.includeClient(), provider)

	private fun <T : DataProvider> buildServer(event: GatherDataEvent, generator: DataGenerator, provider: T): T =
		generator.addProvider(event.includeServer(), provider)
}
